### PR TITLE
Firebase Data Collection:- Add tri states to the data collection api

### PR DESCRIPTION
**Pull request by VinayGuthal**
Add tri state for the dataCollectionsetter in firebase-common

### CODE REVIEW

File: firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java [hunk 1]
```diff
@@ -37,7 +36,7 @@
   private final Context applicationContext;
   private final SharedPreferences sharedPreferences;
   private final Publisher publisher;
-  private final AtomicBoolean dataCollectionDefaultEnabled;
+  private Boolean dataCollectionDefaultEnabled;
```
Comment (vkryachko) [inline firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java:39]: Why is it no longer atomic? do we not care about concurrency anymore?

File: firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java [hunk 2]
```diff
@@ -61,19 +61,28 @@ public boolean isEnabled() {
...
+      if (dataCollectionDefaultEnabled.compareAndSet(!manifestSetting, manifestSetting)) {
+        publisher.publish(
+            new Event<>(
+                DataCollectionDefaultChange.class,
+                new DataCollectionDefaultChange(manifestSetting)));
+      }
+    } else {
+      boolean apiSetting = Boolean.TRUE.equals(enabled);
+      sharedPreferences.edit().putBoolean(DATA_COLLECTION_DEFAULT_ENABLED, apiSetting).apply();
+      if (dataCollectionDefaultEnabled.compareAndSet(!apiSetting, apiSetting)) {
```
Comment (ashwinraghav) [inline firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java:77]: Can be extracted into a function?

File: firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java [hunk 3]
```diff
@@ -61,19 +61,28 @@ public boolean isEnabled() {
...
 
-  public void setEnabled(boolean enabled) {
-    if (dataCollectionDefaultEnabled.compareAndSet(!enabled, enabled)) {
-      sharedPreferences.edit().putBoolean(DATA_COLLECTION_DEFAULT_ENABLED, enabled).apply();
-
-      publisher.publish(
-          new Event<>(DataCollectionDefaultChange.class, new DataCollectionDefaultChange(enabled)));
+  public void setEnabled(Boolean enabled) {
+    if (enabled == null) {
+      sharedPreferences.edit().remove(DATA_COLLECTION_DEFAULT_ENABLED).apply();
```
Comment (ashwinraghav) [inline firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java:66]: Should this also live inside the synchronized block?

File: firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java [hunk 4]
```diff
@@ -61,19 +61,28 @@ public boolean isEnabled() {
...
+      boolean manifestSetting = readManifestDataCollectionEnabled();
+      if (dataCollectionDefaultEnabled.compareAndSet(!manifestSetting, manifestSetting)) {
+        publisher.publish(
+            new Event<>(
+                DataCollectionDefaultChange.class,
+                new DataCollectionDefaultChange(manifestSetting)));
+      }
+    } else {
+      boolean apiSetting = Boolean.TRUE.equals(enabled);
```
Comment (ashwinraghav) [inline firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java:75]: Synchronized block?

File: firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java (hunk 3 above)
Comment (VinayGuthal) [reply firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java:66]: I think shared preferences itself makes sure that it's synchornized

File: firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java (hunk 4 above)
Comment (VinayGuthal) [reply firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java:75]: I think shared preferences itself makes sure that it's synchornized

File: firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java (hunk 2 above)
Comment (VinayGuthal) [reply firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java:77]: The publisher part?

File: firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java (hunk 3 above)
Comment (ashwinraghav) [reply firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java:66]: Sure, you may however run into a situation where one thread sets the shared prefs value and another one broadcasts the message. So they need to be synchronized around the same mutex

File: firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java (hunk 4 above)
Comment (ashwinraghav) [reply firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java:75]: I mean Lines 69-73 and lines 77-82 seem extractable into a reusable function

Comment (VinayGuthal) [reply firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java:75]: yeah makes sense i will do that

File: firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java (hunk 3 above)
Comment (VinayGuthal) [reply firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java:66]: I dont think a synchronized block is needed in this case. Since the method is doing only one write and no read from the shared prefs. And since readManifestDataCollectionEnabled always returns the same value based on the manifest file

Comment (vkryachko) [reply firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java:66]: You *do* need to synchronize here, before your change synchronization was happening with the atomic bool, but since you are checking for null now, a bunch of races become possible, i.e. concurrent threads can enter different branches of this if block and cause mess up the resulting state.

Comment (VinayGuthal) [reply firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java:66]: actually nvm the setEnabled method should be in a sync block..
setEnabled(true) -> Thread1
setEnabled(false) -> Thread2
This might result in the shared preferences having the value of false but the dataCollection having the value of True based on the order of execution

File: firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java [hunk 5]
```diff
@@ -43,15 +43,14 @@ public DataCollectionConfigStorage(
...
     this.applicationContext = directBootSafe(applicationContext);
     this.sharedPreferences =
-        applicationContext.getSharedPreferences(
+        this.applicationContext.getSharedPreferences(
             FIREBASE_APP_PREFS + persistenceKey, Context.MODE_PRIVATE);
     this.publisher = publisher;
     this.dataCollectionDefaultEnabled = new AtomicBoolean(readAutoDataCollectionEnabled());
   }
 
   private static Context directBootSafe(Context applicationContext) {
```
Comment (vkryachko) [inline firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java:52]: You may want to rename this to `deviceProtectedContext` since you changed the logic inside

File: firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java [hunk 6]
```diff
@@ -61,19 +60,26 @@ public boolean isEnabled() {
     return dataCollectionDefaultEnabled.get();
   }
 
-  public void setEnabled(boolean enabled) {
+  private synchronized void updateDataCollectionDefaultEnabled(boolean enabled) {
     if (dataCollectionDefaultEnabled.compareAndSet(!enabled, enabled)) {
```
Comment (vkryachko) [inline firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java:64]: Since the method is now synchronized there is no point for the bool to be atomic. Also very important to make isEnabled() synchronized as well.

File: firebase-common/data-collection-tests/src/test/java/com/google/firebase/DataCollectionTestUtil.java [hunk 7]
```diff
@@ -52,10 +52,17 @@ static SharedPreferences getSharedPreferences() {
             Context.MODE_PRIVATE);
```
Comment (vkryachko) [inline firebase-common/data-collection-tests/src/test/java/com/google/firebase/DataCollectionTestUtil.java:52]: How does this test work if it's using user credential encrypted context?(i.e. not the direct boot one?)

Comment (VinayGuthal) [reply firebase-common/data-collection-tests/src/test/java/com/google/firebase/DataCollectionTestUtil.java:52]: Its because of the config sdk=19

File: firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java (hunk 5 above)
Comment (vkryachko) [reply firebase-common/src/main/java/com/google/firebase/internal/DataCollectionConfigStorage.java:51]: Actually I think this method is even redundant and you can just use `ContextCompat.createDeviceProtectedStorageContext()` directly as it does the same version check.

File: firebase-crashlytics/src/androidTest/java/com/google/firebase/crashlytics/internal/common/CrashlyticsControllerTest.java [hunk 8]
```diff
@@ -842,9 +842,8 @@ public void testUploadDisabledThenEnabled() throws Exception {
     Task<Void> task = controller.submitAllReports(1.0f, testSettingsDataProvider.getAppSettings());
 
     arbiter.setCrashlyticsDataCollectionEnabled(true);
-
+    
     await(task);
-
```
Comment (vkryachko) [inline firebase-crashlytics/src/androidTest/java/com/google/firebase/crashlytics/internal/common/CrashlyticsControllerTest.java:847]: pls revert this change, lgtm otherwise
